### DISCUSSION TITLE
ConvertTo-WDACPolicy cmdlet Publisher type instead of Hashes

### DISCUSSION THREAD

**Discussion by a-shina** [HotCakeX/Harden-Windows-Security#259] [Q&A] [answered]
Hello,

First off, I would like to say thank you for creating the WDACConfig module, it is super helpful and easy to use!

Unfortunately, I am experiencing this issue where the supplemental policy created through MDE Advanced Hunting is using only hashes for allow rules instead of Publisher type even though the files are signed.

I have also watched your YouTube video on how to use WDACConfig module with Microsoft Defender for Endpoint (MDE) Advanced Hunting results to create Application Control policies and followed the steps as applicable.

Is there any configuration I need to adjust or additional steps I should take to ensure all publisher fields are included when parsing from MDE Advanced Hunting events?

Thank you for your support!

**Comment by HotCakeX** [comment]
Hi,
my pleasure, glad you find it useful,
Can you share some parts of the CSV file with me so i can investigate further? You can send through email or on Teams: spynetgirl@outlook.com

The module first tries to create FilePublisher, if necessary details aren't available then attempts to create Publisher and lastly tries Hash level rules. If the CSV contains file's signatures (Publisher Name, Publisher TBS Hash, Issuer Name, Issuer TBS Hash) then Publisher level should be used.

**Reply by a-shina** [reply]
Thank you for your prompt response! Just sent you an email.

**Reply by HotCakeX** [reply]
Thank you! i checked your CSV file and there is absolutely no file signature info in there!
You can verify it too, just open the CSV file in VS Code, Excel or even Notepad, then search for these keywords

`PublisherTBSHash`
`PublisherName`
`IssuerName`
`IssuerTBSHash`

or even these

`tbs`
`issuer`
`publisher`

You will find nothing. Those are the necessary fields that MDE Advanced Hunting generates for signed files. They are also included in the local Code Integrity event logs.

So I'm thinking either the files are all unsigned or there was a problem exporting the MDE logs to the CSV. I inspected the module's workflows and it's detecting all files as unsigned based on the input CSV file as expected.

Please refer to this [document](https://github.com/HotCakeX/Harden-Windows-Security/wiki/How-to-Use-Microsoft-Defender-for-Endpoint-Advanced-Hunting-With-WDAC-App-Control) and try exporting them again.

This KQL query should include all of the info for signed/unsigned files

```kql
DeviceEvents
| where ActionType startswith "AppControlCodeIntegrity"
   or ActionType startswith "AppControlCIScriptBlocked"
   or ActionType startswith "AppControlCIScriptAudited"
```

**Reply by a-shina** [reply]
Thank you!  There appears to be a deeper issue going on here.  For some reason, when parsing event logs directly from the system event viewer it generates all publisher fields but when I use MDE AH these fields are missing!
 
Let's take row 5 from the csv file as an example:
FileName: SnagitEditor.exe
InitiatingProcessVersionInfoCompanyName: TechSmith Corporation

We would expect to see the publisher info in the AdditionalFields from the query right?  But when we look at the corresponding event from the csv file, it does not appear that the publisher fields are generating.

The screenshot below is using the WDAC Wizard to parse event logs from the system event viewer to policy.  This is how I would expect the publisher info to show up from the MDE AH query.
![WDAC-Wizard-SnagitEditor](https://github.com/HotCakeX/Harden-Windows-Security/assets/55485702/4cd80405-c8e9-49b3-809b-c09f0e373eb5)

I also ran this query again to verify and filtered specifically for this file mentioned above and all 4 events in the results do not have publisher info:
```
DeviceEvents
| where DeviceName startswith "sdashina"
| where ActionType startswith "AppControlCodeIntegrity"
    or ActionType startswith "AppControlCIScriptBlocked"
    or ActionType startswith "AppControlCIScriptAudited"
| where FileName contains "SnagitEditor.exe"
```

Any idea what could be causing this?  I will also open a support ticket with Microsoft since this looks like an issue on their end but I would love to hear your input in case you have any suggestions!

Thank you again,
Adrian

**Reply by HotCakeX** [reply]
You're on the right track, the info you found is definitely useful and can help in the support ticket conversations.

What I suspect is the type of Audit policy deployed on the system that is generating these events. Can you try again the entire process?
Remove any audit mode policy from the system, and this time *only* deploy the AllowMicrosoft policy in audit mode. You can use WDAC Wizard for it.

I wouldn't recommend using the SignedAndReputable policy because its decisions are based on the ISG and it can change based on the global signal intelligence. Once you do that, grab the new logs from MDE AH and see if the signature info is included in the CSV file.

**Reply by a-shina** [reply]
I will definitely give that a shot -- thank you very much!

**Reply by HotCakeX** [reply]
No problem, yw! :)

**Reply by a-shina** [reply]
Hi again @HotCakeX, hope all is well!

I was wondering if there is a way using the ConvertTo-WDACPolicy cmdlet when creating an allow rule to only have it create based on Issuer & Publisher and omit File Attribute?  This option is available using the WDAC Wizard and it would be so helpful to have it on your WDACConfig module as well.

Example from the Wizard:
![image](https://github.com/user-attachments/assets/c02dc2b4-ecbf-412b-83d8-f07090b9ab5a)
![image](https://github.com/user-attachments/assets/5f81e37f-7b48-475b-a56b-ebdf2cb1e3bb)

Shown above, when creating an allow rule for DellPair.exe - I am able to create it based on Issuer & Publisher only.

Examples of xml generated from:

1. WDAC Wizard:
![image](https://github.com/user-attachments/assets/8853f9eb-2970-48da-adfc-4c0d6d516f66)

2. WDACConfig:
![image](https://github.com/user-attachments/assets/14e85d6a-2ef8-4236-b131-92e96fd8b59e)
![image](https://github.com/user-attachments/assets/01d0464c-0ed2-455b-917e-b29b4fd93c8e)


The reason I prefer this option is that any other file or executable under DigiCert EV Code Signing CA (SHA2) and IndiLogic LLC would be allowed using only one allow rule. 

Thanks,
Adrian

**Reply by HotCakeX** [reply, marked as answer]
Hi, when i designed that function, i made it to use the [most secure levels](https://github.com/HotCakeX/Harden-Windows-Security/wiki/WDAC-Rule-Levels-Comparison-and-Guide) as much as possible. So it first tries to use FilePublisher level, if the logs can't satisfy the requirements for it, then it falls back to Publisher level, if that can't be satisfied it tries to use LeafCertificate, and finally it will create Hash rules.

I'm beginning to realize people (including you) want to use Publisher level primarily, especially in enterprise environments, so i will offer level selection in the command. The logics are already there, just need to take care of some minor technicalities to make it happen. Next version of the WDACConfig module will most likely offer it.

**Reply by a-shina** [reply]
Awesome!  Thank you for considering this enhancement. Having this level of granularity will be super nice. Do you have an estimated release date for the next update?  I'm looking forward to it!

**Reply by HotCakeX** [reply]
@a-shina Just released the update
https://github.com/HotCakeX/Harden-Windows-Security/releases/tag/WDACConfigv0.4.3

sorry couldn't give an estimation

**Reply by a-shina** [reply]
@HotCakeX, thank you so much for the swift release! I’m excited to start using the new features. Much appreciated :)

**Reply by HotCakeX** [reply]
You're welcome ^^
